modules/auxiliary/gather/shodan_search.rb

```diff
@@ -110,42 +110,46 @@ def run
     # create our Shodan request parameters
     query = datastore['QUERY']
     apikey = datastore['SHODAN_APIKEY']
-    page = 1
     maxpage = datastore['MAXPAGE']
 
     # results gets our results from shodan_query
     results = []
-    results[page] = shodan_query(query, apikey, page)
+    results[0] = shodan_query(apikey, query, 1)
 
-    if results[page]['total'].nil? || results[page]['total'] == 0
+    if results[0]['total'].nil? || results[0]['total'] == 0
       msg = "No results."
-      if results[page]['error'].to_s.length > 0
-        msg << " Error: #{results[page]['error']}"
+      if results[0]['error'].to_s.length > 0
+        msg << " Error: #{results[0]['error']}"
       end
       print_error(msg)
       return
     end
 
     # Determine page count based on total results
-    if results[page]['total'] % 100 == 0
-      tpages = results[page]['total'] / 100
+    if results[0]['total'] % 100 == 0
+      tpages = results[0]['total'] / 100
     else
-      tpages = results[page]['total'] / 100 + 1
-      maxpage = tpages if datastore['MAXPAGE'] > tpages
+      tpages = results[0]['total'] / 100 + 1
     end
+    maxpage = tpages if datastore['MAXPAGE'] > tpages
 
     # start printing out our query statistics
-    print_status("Total: #{results[page]['total']} on #{tpages} " +
+    print_status("Total: #{results[0]['total']} on #{tpages} " +
       "pages. Showing: #{maxpage} page(s)")
 
     # If search results greater than 100, loop & get all results
     print_status('Collecting data, please wait...')
-    if results[page]['total'] > 100
-      page += 1
-      while page <= maxpage
-        break if page > datastore['MAXPAGE']
-        results[page] = shodan_query(query, apikey, page)
-        page += 1
+    p = 1
+    if results[0]['total'] > 100
+      page = 2
+      while p < maxpage
+        results[p] = shodan_query(apikey, query, page)
+        if results[p]['matches'].nil?
+          next
+        else
+          p += 1
+          page += 1
+        end
       end
     end
```

Comment:
Since at this point, `maxpages` has been set to the lower value between the user-specified option and the total pages, this could be safely reduced to:

```suggestion
    maxpages.times do |page|
      page_results = shodan_query(apikey, query, page + 1)
      break if page_results['matches'].nil?

      results[page] = page_results
    end
```

Drop the `page + 1` in the `shodan_query` API call if you implemented the other feedback to treat the parameter as a page index.